src/Core/Lib/ConcurrentObservableCollection.cs

```diff
@@ -0,0 +1,109 @@
+#region License
+/* 
+ * Copyright (C) 1999-2021 Pavel Tomin.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Text;
+
+namespace Reko.Core.Lib
+{
+    /// <summary>
+    /// Represents thread-safe collection that provides notifications
+    /// when items get added, removed, or when the whole list is refreshed.
+    /// This is lock-based wrapper of
+    /// <see cref="Reko.Core.Lib.ObservableRangeCollection{T}"/>
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ConcurrentObservableCollection<T> : IEnumerable<T>
+    {
+        public event NotifyCollectionChangedEventHandler? CollectionChanged
+        {
+            add
+            {
+                lock (collection)
```

Comment:
A remark, perhaps worth adding to the source: all these locks will make the collection thread safe but will also cause performance to drop if there is contention on the lock.

Reply:
Done in bf90ce3